plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -761,6 +762,10 @@ private ObjectMetadata getS3ObjectMetadata(Path path, String bucketName, String
                         }
                         catch (RuntimeException e) {
                             STATS.newGetMetadataError();
+                            if (e instanceof AWSSecurityTokenServiceException
```

Comment:
I believe the AWS SDK already retries on throttling errors. Take a look at how the S3 client is configured in TrinoS3Filesystem to get an idea.

Reply:
Indeed - STS specific error codes are documented at https://docs.aws.amazon.com/STS/latest/APIReference/CommonErrors.html.

What a weird choice. Can you add this context to the commit message (and maybe a well-placed code comment - it's not obvious and useful context).

Reply:
@hashhar Added some more context to my commit, https://github.com/trinodb/trino/pull/10535/commits/d9f6eaccc9ad73546ae51ced13a98b69a3b53f9e.

What's even stranger is that the error code returned via the API does not match their documentation which is why in the commit I check for `Throttling` as opposed to what is in the documentation (`ThrottlingException`).

Reply:
@pettyjamesm is it expected for Glue to use HTTP 4xx for throttling? Or does the 4xx happen only when Glue really doesn't want the automatic retry from AWS SDK to kick in?

Any opinions on how we should handle STS throttling errors?

Reply:
The throttling exception in the stacktrace above is coming from the STS service as part of the `STSAssumeRoleSessionCredentialsProvider.getCredentials` call _before_ the service (S3 / Glue) is called in order to get usable credentials. So it's not Glue or S3 that's actually throwing this exception, it's the pre-request logic trying to get usable credentials to issue the request with.

It's a completely separate AWS client that may need some configuration knobs tuned to mitigate this issue- depending on how the `STSAssumeRoleSessionCredentialsProvider` was created in the first place. For instance if the `ClientConfiguration#throttleRetries` property is enabled (which I believe it is by default) then the client could go into a fail-fast circuit breaker mode and RetryPolicy outside of the AWS client won't have any effect. When you run up against throttling at the credential provider layer like this it may be worth figuring out whether there's some bug causing too many assume role calls being invoked (like maybe separate credentials providers are being created each calling STS independently?).

Reply:
I added a bit more clarity here as well https://github.com/trinodb/trino/pull/10535#discussion_r783513691. Yes @pettyjamesm, we utilize single tenant AWS glue catalogs for our customers. This essentially happens every time we re-start the cluster which we do often due to turning off the clusters when no queries are present.

On our largest clusters of at least a 100 workers, it takes a large number of client retries usually between 20-50 to finally get all of the workers with their credentials set correctly. Where handling it in Trino with the default retry logic seems more fitting since it is a retry-able error according to AWS.